ffi/tests/ctests/src/tests.c

```diff
@@ -0,0 +1,219 @@
+	field = wirefilter_string("ssl");
+	wirefilter_add_type_field_to_scheme(scheme, field, WIREFILTER_TYPE_BOOL);
+	field = wirefilter_string("tcp.port");
+	wirefilter_add_type_field_to_scheme(scheme, field, WIREFILTER_TYPE_INT);
+
+	filter_str = wirefilter_string("tcp.port == 80");
+	wirefilter_parsing_result_t result = wirefilter_parse_filter(scheme, filter_str);
+	assert(result.success == true, "could not parse good filter");
+	assert(result.ok.ast != NULL, "could not parse good filter");
+
+	wirefilter_filter_t *filter = wirefilter_compile_filter(result.ok.ast);
+	assert(filter != NULL, "could not compile filter");
+
+	wirefilter_free_compiled_filter(filter);
+
+	wirefilter_free_scheme(scheme);
+}
+
+void wirefilter_ffi_ctest_create_execution_context() {
+	wirefilter_scheme_t *scheme = wirefilter_create_scheme();
+	assert(scheme != NULL, "could not create scheme");
+
+	wirefilter_execution_context_t *exec_ctx = wirefilter_create_execution_context(scheme);
+	assert(exec_ctx != NULL, "could not create execution context");
+
+	wirefilter_free_execution_context(exec_ctx);
+
+	wirefilter_free_scheme(scheme);
+}
+
+void wirefilter_ffi_ctest_add_values_to_execution_context() {
+	wirefilter_scheme_t *scheme = wirefilter_create_scheme();
+	assert(scheme != NULL, "could not create scheme");
+
+	wirefilter_externally_allocated_str_t field;
+
+	field = wirefilter_string("http.host");
+	wirefilter_add_type_field_to_scheme(scheme, field, WIREFILTER_TYPE_BYTES);
+	field = wirefilter_string("ip.addr");
+	wirefilter_add_type_field_to_scheme(scheme, field, WIREFILTER_TYPE_IP);
+	field = wirefilter_string("ssl");
+	wirefilter_add_type_field_to_scheme(scheme, field, WIREFILTER_TYPE_BOOL);
+	field = wirefilter_string("tcp.port");
+	wirefilter_add_type_field_to_scheme(scheme, field, WIREFILTER_TYPE_INT);
+
+	wirefilter_execution_context_t *exec_ctx = wirefilter_create_execution_context(scheme);
+	assert(exec_ctx != NULL, "could not create execution context");
+
+	field = wirefilter_string("http.host");
+	wirefilter_externally_allocated_byte_arr_t http_host;
+	http_host.data = (unsigned char *)"www.cloudflare.com";
+	http_host.length = strlen((char *)http_host.data);
+	wirefilter_add_bytes_value_to_execution_context(exec_ctx, field, http_host);
+
+	field = wirefilter_string("ip.addr");
+	uint8_t ip_addr[4] = {192, 168, 0, 1};
+	wirefilter_add_ipv4_value_to_execution_context(exec_ctx, field, ip_addr);
+
+	field = wirefilter_string("ssl");
+	wirefilter_add_bool_value_to_execution_context(exec_ctx, field, false);
+
+	field = wirefilter_string("tcp.port");
+	wirefilter_add_int_value_to_execution_context(exec_ctx, field, 80);
+
+	wirefilter_free_execution_context(exec_ctx);
+
+	wirefilter_free_scheme(scheme);
+}
+
+void wirefilter_ffi_ctest_match_filter() {
+	wirefilter_scheme_t *scheme = wirefilter_create_scheme();
+	assert(scheme != NULL, "could not create scheme");
+
+	wirefilter_externally_allocated_str_t field;
+
+	field = wirefilter_string("http.host");
```

Comment:
Maybe inline these as function params instead of using separate variable?

Reply:
Personally i kind of prefer the current code, i find it more readable but I don't really have a strong opinion about this.

Reply:
It slightly worries me that, when adding new fields, it's easy to forget to add `field = ...` line and just call a following function with a previous value, whereas with an inline expression you can't miss it.

Reply:
Fixed :)